lib/iris/cube.py

```diff
@@ -4075,98 +4075,65 @@ def aggregated_by(
         data_shape = list(self.shape + aggregator.aggregate_shape(**kwargs))
         data_shape[dimension_to_groupby] = len(groupby)
 
-        # Aggregate the group-by data.
+        # Choose appropriate data and functions for data aggregation.
         if aggregator.lazy_func is not None and self.has_lazy_data():
-            front_slice = (slice(None, None),) * dimension_to_groupby
-            back_slice = (slice(None, None),) * (
-                len(data_shape) - dimension_to_groupby - 1
-            )
+            stack = da.stack
+            input_data = self.lazy_data()
+            agg_method = aggregator.lazy_aggregate
+        else:
+            input_data = self.data
+            # Note numpy.stack does not preserve masks.
+            stack = ma.stack if ma.isMaskedArray(input_data) else np.stack
+            agg_method = aggregator.aggregate
+
+        # Create data and weights slices.
+        front_slice = (slice(None),) * dimension_to_groupby
+        back_slice = (slice(None),) * (
+            len(data_shape) - dimension_to_groupby - 1
+        )
 
-            # Create cube and weights slices
-            groupby_subcubes = map(
-                lambda groupby_slice: self[
+        groupby_subarrs = map(
+            lambda groupby_slice: iris.util._slice_data_with_keys(
```

Comment:
Note that `_slice_data_with_keys` is used by [`Cube.__getitem__`](https://github.com/SciTools/iris/blob/de911ce55da50172b51c662c44a7d1adff148dca/lib/iris/cube.py#L2541), so this is just a shortcut to what was already happening.  Should be a little more efficient.

Reply:
> Should be a little more efficient

Slicing just the array rather than the whole cube seems to account for the entire speedup noted at https://github.com/SciTools/iris/pull/4970#issuecomment-1263663273